### PR TITLE
Make estimator return block hashes instead of full block 

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

Some calls to the estimator don't require the full BlockMessage
to be returned. Thus we should only return the hashes and
let the caller fetch the BlockMessage based on those hashes
if they need to.

This also removes the BlockStore effect from the Estimator.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

https://rchain.atlassian.net/browse/RCHAIN-3069

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by KentShikama** [comment]
bors try

**Comment by KentShikama** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/Estimator.scala [hunk 1]
```diff
@@ -128,17 +121,14 @@ object Estimator {
...
-                    }
-                  } yield result
+                  blockDag.lookup(childHash).map {
+                    case Some(childBlock)
+                        if childBlock.parents.size > 1 && childBlock.sender != validator =>
+                      val currScore       = acc2.getOrElse(childHash, 0L)
+                      val validatorWeight = childBlock.weightMap.getOrElse(validator, 0L)
+                      acc2.updated(childHash, currScore + validatorWeight)
+                    case _ => acc2
+                  }
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/Estimator.scala]: My head hurts when I try to understand this function. Please make it more readable.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Estimator.scala]: The change is NFC. Hopefully the after version is better than the before. If not I'll just revert this entirely.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Estimator.scala]: I've also just considered removing this part. It was initially an idea to make the estimator a bit more fair but for the amount of thinking required verify its correctness I don't think it is worth it.

File: casper/src/main/scala/coop/rchain/casper/Estimator.scala [hunk 2]
```diff
@@ -190,17 +180,16 @@ object Estimator {
...
+    blockDag.children(b).map {
+      case Some(children) =>
+        val scoredChildren = children.filter(scores.contains)
+        if (scoredChildren.nonEmpty) {
+          scoredChildren.toList
+        } else {
+          List(b)
+        }
+      case None => List(b)
+    }
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/Estimator.scala]: Trying to understand what happens here I got to the following refactoring which I find more readable (but I don't have too strong opinions):

```scala
 private  def toNonEmptyList[T](elements  :Set[T]) : Option[List[T]] =
    if (elements.isEmpty) None
    else Some(elements.toList)

  private def replaceBlockHashWithChildren[F[_]: Monad](
      b: BlockHash,
      blockDag: BlockDagRepresentation[F],
      scores: Map[BlockHash, Long]
  ): F[List[BlockHash]] =
    blockDag.children(b)
      .map (
       _.flatMap { children => toNonEmptyList(children.filter(scores.contains))}
        .getOrElse(List(b))
      )
```

File: casper/src/main/scala/coop/rchain/casper/Estimator.scala (hunk 1 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Estimator.scala]: On further thought, I'm going to remove this indeed.

File: casper/src/main/scala/coop/rchain/casper/Estimator.scala (hunk 2 above)
Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/Estimator.scala]: What about

```
  private def toNonEmptyList[T](elements: Set[T]) : Option[List[T]] =
    if (elements.isEmpty) None
    else Some(elements.toList)

  private def replaceBlockHashWithChildren[F[_]: Monad](
                                                         b: BlockHash,
                                                         blockDag: BlockDagRepresentation[F],
                                                         scores: Map[BlockHash, Long]
                                                       ): F[List[BlockHash]] =
    blockDag.children(b).map {
      case Some(children) =>
        val scoredChildren = children.filter(scores.contains)
        toNonEmptyList(scoredChildren).getOrElse(List(b))
      case None =>
        List(b)
    }
```

**Comment by KentShikama** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala [hunk 3]
```diff
@@ -320,21 +320,27 @@ object ProtoUtil {
...
-            acc.pure[F]
-          )
-      }
-      .map(_.reverse)
+    for {
+      blocks <- blockHashes.toList.traverse(hash => ProtoUtil.unsafeGetBlock[F](hash))
+      result <- blocks
+                 .foldM(List.empty[BlockMessage]) {
+                   case (acc, b) =>
+                     acc.forallM(nonConflicting(b)).map { isNonConflictingWithAll =>
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: the previous version seemed easier to read. why the change?

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: I thought this version was easier to read; the Monad[F].ifM syntax just doesn't read cleanly to me.

Comment (odeac) [reply casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: I agree with Dom that the previuos was cleaner. And even if ifM is difficult to read, because of the fact that it's a standard operator it is easy to learn once and then expect the same behaviour everywhere

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala]: Ok I'll revert to using ifM.

**Comment by KentShikama** [comment]
bors try
